apicurio-registry/pkg/controller/apicurioregistry/apicurioregistry_controller.go

```diff
@@ -0,0 +1,196 @@
+
+// ReconcileApicurioRegistry reconciles a ApicurioRegistry object
+type ReconcileApicurioRegistry struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client  client.Client
+	scheme  *runtime.Scheme
+	created int32
+}
+
+// Reconcile reads that state of the cluster for a ApicurioRegistry object and makes changes based on the state read
+// and what is in the ApicurioRegistry.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileApicurioRegistry) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling PodSet")
+
+	// Fetch the PodSet instance
+	podSet := &registryv1alpha1.ApicurioRegistry{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, podSet)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// List all pods owned by this PodSet instance
+	lbls := labels.Set{
+		"app": podSet.Name,
+	}
+	existingPods := &corev1.PodList{}
+	err = r.client.List(
+		context.TODO(),
+		&client.ListOptions{
+			Namespace:     request.Namespace,
+			LabelSelector: labels.SelectorFromSet(lbls),
+		},
+		existingPods)
+	if err != nil {
+		reqLogger.Error(err, "failed to list existing pods in the podSet")
+		return reconcile.Result{}, err
+	}
+	existingPodNames := []string{}
+	// Count the pods that are pending or running as available
+	for _, pod := range existingPods.Items {
+		if pod.GetObjectMeta().GetDeletionTimestamp() != nil {
+			continue
+		}
+		if pod.Status.Phase == corev1.PodPending || pod.Status.Phase == corev1.PodRunning {
+			existingPodNames = append(existingPodNames, pod.GetObjectMeta().GetName())
+		}
+	}
+
+	reqLogger.Info("Checking podset", "expected replicas", podSet.Spec.Replicas, "Pod.Names", existingPodNames)
+	// Update the status if necessary
+	status := registryv1alpha1.ApicurioRegistryStatus{
+		Replicas: int32(len(existingPodNames)),
+		PodNames: existingPodNames,
+	}
+	if !reflect.DeepEqual(podSet.Status, status) {
+		podSet.Status = status
+		err := r.client.Status().Update(context.TODO(), podSet)
+		if err != nil {
+			reqLogger.Error(err, "failed to update the podSet")
+			return reconcile.Result{}, err
+		}
```

Comment:
Two comments:
* Requeue when done with status reconciling
* Leave status reconciliation as the last step in the whole reconciliation process.